Marlin/Marlin_main.cpp

```diff
@@ -499,27 +497,28 @@ static bool drain_queued_commands_P() {
  */
 void enqueuecommands_P(const char* pgcode) {
   queued_commands_P = pgcode;
```

Comment:
Aha, so "flowcharting" does lead to the same obvious conclusion. You can't inject directly into the buffer and leave it half-done between loops — not if you also want to be able to jump in anytime and do command injection. Because of the way that serial input is being processed, any code that calls enqueuecommand() will break the serial buffer, and the next command will contain garbage characters from an earlier command (16 commands ago for a ring buffer of size 16).

There is a simple solution that doesn't require any complex buffer, and that is to simply add a static serial line buffer for the serial input! Then it can just call `enqueuecommand` to copy the whole command into the buffer, never leaving the ring buffer in a blocked state. All code that looks at `serial_count != 0` can forget about that condition. This means also that after the serial input is processed in `get_command` the code can simply fall through to the SD file processing. Thus an incomplete line sitting on the serial input won't block SD file processing. Even though the extra string copy costs some cycles, small efficiency improvements like these start to make up for it.